dbms/src/Flash/Mpp/MPPTunnel.cpp

```diff
@@ -98,24 +97,26 @@ MPPTunnel::~MPPTunnel()
     LOG_FMT_TRACE(log, "destructed tunnel obj!");
 }
 
-void MPPTunnel::finishSendQueue()
+void MPPTunnel::finishSendQueue(bool drain)
 {
-    bool flag = send_queue->finish();
-    if (flag && mode == TunnelSenderMode::ASYNC_GRPC)
+    if (tunnel_sender)
     {
-        async_tunnel_sender->tryFlushOne();
+        if (drain)
+            tunnel_sender->finishAndDrain();
+        else
+            tunnel_sender->finish();
     }
 }
 
 /// exit abnormally, such as being cancelled.
 void MPPTunnel::close(const String & reason)
 {
     SCOPE_EXIT({
-        // ensure the tracked memory is released and udpated before memotry tracker(in ProcListEntry) is released
-        send_queue->finishAndDrain(); // drain the send_queue when close
+        // ensure the tracked memory is released and updated before memory tracker(in ProcListEntry) is released
+        finishSendQueue(true); // drain the send_queue when close
```

Comment:
The original one didn't triger a tryFlushOne()/kick, however, the new one will triger. Is it expected?

Reply:
It's safe to call `kick` function multiple times. But calling `kick` function here is indeed useless in current implementation. I added a `kick` function in `GRPCSendQueue::finishAndDrain` just for semantic consistency.